docs(DHT): update dht README (#2350)
streamr-dev/network: packages/dht/README.md

# @streamr/dht

> A connectionful Kademlia based P2P distributed hash table (DHT). Implements an in-memory key-value data store where multiple creators of of data can list values behind a single key. The library is entirely browser compatible.

Reviewer: ```suggestion
A connectionful Kademlia based P2P distributed hash table (DHT). Implements an in-memory key-value data store where multiple creators of data store values behind a single key. The library is entirely browser compatible.
```

> All communication between peers utilizes protobuf, transported with the help of the custom `@streamr/proto-rpc` library.

Reviewer: ```suggestion
All communication between peers utilizes [protobuf](https://protobuf.dev/), transported with the help of the custom `@streamr/proto-rpc` library.
```

Connections in the DHT are established using WebSocket or WebRTC connections. The nodes will decide internally based on peers' connectivity information which connection types to use.

> The DHT also provides an interface for sending messages from peer to peer with the `DhtNode#send` function. The sent message is routed over the network to the target node's id.

Reviewer: ```suggestion
The DHT also provides an interface for sending messages from peer to peer with the `DhtNode#send` function. The sent message is routed over the network to the target node.
```

Could also also briefly describe how the network topology is formed? I.e. nodes which have closest kademlia distance are neighbors, and how `randomPeers` is related to that? Are we always connected to all neighbors, or other nodes too?

Author: Not necessarily important to detail that here. Technical details can be added in a separate section in the future

## Running a node

Running a bare DhtNode in the Streamr Network is heavily discouraged as it requires advanced configuration. However, if you do wish to run a node here are some recommended configurations.

> (the following values are examples for working production or test values check the streamr-client's default network configuration)

Reviewer: ```suggestion
(the following values are examples; for working production or test values check the streamr-client's default network configuration)
```

```js
const DhtNode = new DhtNode({
  // Add a list of known entry points to the network
  entryPoints: [{
    kademliaId: new Uint8Array([1, 2, 3])
  }], 
  // a list of STUN and TURN servers. Critically important when opening webrtc conenctions behind NATs.
  iceServers: [{
    url: "stun.l.google.com",
    port: 19302
  }],

>   // A range of ports that the node will attempt to start a WebSocket server in. If you wish to use an exact port give equal values to min and max.

Reviewer: It is also okay to start a node without a WebSocket server? What are consequences of that?

  websocketPortRange: {
    min: 30000,
    max: 30500
  },
  ...
})
```

## WebSocket Server TLS configuration

Setting up TLS is important 

There two ways to configure TLS for a DhtNode's WS server.

### Auto-certification

When starting a node configure it to automatically fetch TLS certificates for a randomly generated domain name as such:

```js
const DhtNode = new DhtNode({
  ...

>   websocketServerEnableTls: true,

Reviewer: Not related to this PR: maybe `autoCertified* `config options could be moved into the `tlsCertificate` block in the future?

Author: yeah at some point configuration related to specific features could be given their specific objects

  ...
})
```

### Custom TLS configuration

If you have your own domain name and/or TLS certificates you can configure the node to use them as follows:

```js
const DhtNode = new DhtNode({
  tlsCertificate: {
    certFileName: 'path/to/file'
    privateKeyFileName: 'path/to/file'
  },
  websocketHost: 'custom-domain.com' // If using a custom domain name give it here to ensure that connectivity checking is correctly pointed to the server 
})
```

### Setting up a custom DHT network

To setup a custom DHT network outside the Streamr Networks Mumbai or Polygon environments, you simply need to set up one or multiple entry point DhtNodes to the network and point all newly joining DhtNodes to them. It is also a good idea to configure the entry points to know each other. This makes it possible to restart them without causing network partitioning.

## Development

### local testing and develpment setup

When running a node for testing or develpment a few configurations are important.

```js
const DhtNode = new DhtNode({
  entryPoints: [], // Point this to your local dev entry point!
  iceServers: [], // Keep empty to ensure webrtc connections are local

>   webrtcAllowPrivateAddresses: true, // Make sure that the value is true in local develpment

Reviewer: What this does in practice? Allows nodes to connect directly to each others using LAN ip addresses or something?

Author: Yes, if this is false the nodes will filter out private/local ip addresses
